integration/compose/src/main/java/com/bumptech/glide/integration/compose/GlidePainter.kt

```diff
@@ -36,54 +45,345 @@ import kotlinx.coroutines.SupervisorJob
 import kotlinx.coroutines.job
 import kotlinx.coroutines.launch
 import kotlinx.coroutines.plus
+import java.lang.IllegalStateException
+
+/**
+ * Exposes a [Painter] and the painter's [GlidePainterState] so that they can be composed with other
+ * Composables like [androidx.compose.foundation.Image].
+ *
+ * You can use this class to display custom placeholders and/or animations. For example to crossfade
+ * from a placeholder to a real image, you might do something like this:
+ *
+ * ```
+ * val (painter, state) = rememberGlidePainter(model = item.uri) {
+ *   it.override(targetImageSize)
+ * }
+ * when (state.animate()) {
+ *  AnimationState.Loading, AnimationState.Animate -> {
+ *    Crossfade(state == AnimationState.Loading) {
+ *      if (it) {
+ *        Image(
+ *          painterResource(android.R.drawable.star_big_on),
+ *          contentDescription,
+ *          modifier,
+ *        )
+ *      } else {
+ *        Image(painter, contentDescription, modifier)
+ *      }
+ *    }
+ *  }
+ *  AnimationState.Success -> {
+ *    Image(painter, contentDescription, modifier)
+ *  }
+ *  AnimationState.Failed -> {
+ *    Image(
+ *      painterResource(android.R.drawable.star_big_off),
+ *      contentDescription,
+ *      modifier,
+ *    )
+ *  }
+ * }
+ * ```
+ *
+ * [animate] is a helper that uses [AnimationState.Animate] to indicate that an
+ * animation should be performed because the [DataSource] is not [DataSource.MEMORY_CACHE], or
+ * [AnimationState.Success] to indicate that the animation should be skipped. It's derived from
+ * [GlidePainterState] so you can implement your own version if your requirements differ, or skip it
+ * entirely.
+ *
+ * Unlike [GlideImage], the painter cannot automatically determine its size. If no override size
+ * (by calling [RequestBuilder.override] in [requestBuilderTransform]) is provided, the painter will
+ * wait until it is drawn to start the image load. This can lead to the image load never starting.
+ * To ensure the painter is always able to determine a size and start the image load:
+ *
+ * 1. Use the other variant of [rememberGlidePainter] by providing a [Modifier] and using the
+ * [GlidePainterStateAndModifier.modifier] returned by that method.
+ * 2. Ensure that the painter is always drawn (unlike in the example above) in another Composable
+ * (e.g. [androidx.compose.foundation.Image] with a reasonable [Modifier].
+ * 3. Using [requestBuilderTransform] to apply a specific [RequestBuilder.override] size to the
+ * Glide request directly. If you choose this option, try to pick an override size based on the area
+ * you want to display the image in or at least the screen size rather than using
+ * [com.bumptech.glide.request.target.Target.SIZE_ORIGINAL].
+ */
+@ExperimentalGlideComposeApi
+@OptIn(InternalGlideApi::class)
+@Composable
+public fun rememberGlidePainter(
+  model: Any?,
+  requestBuilderTransform: RequestBuilderTransform<Drawable> = { it },
+): Pair<Painter, GlidePainterState> {
```

Comment:
This API is a bit unusual for Compose code. Typically a `remember*` method will return the direct mutable state object, rather than something wrapper in another type such as a `Pair`. For example, I'd expect that this would return a `GlidePainter` object instead. Might be good to create a class like that?

Reply:
I think the disadvantage is that you expose a broader API surface, you get all of GlidePainter when all you need is a state.

I could extract an interface to avoid that:
[
```kotlin 
GlidePainter extends Painter(public val state: GlidePainterState) { }

val painter = rememberGlidePainter(model)
match (painter.state) {
...
}
```

It feels slightly less nice to return a Painter interface that's just Painter + State? That is the way that Coil does this though.

Reply:
Done